### ISSUE TITLE
[Classification] How to obtain TOP1 name and scores

### ISSUE THREAD

**Issue by monkeycc** [ultralytics/ultralytics#2286]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question
probs = result.probs  # Class probabilities for classification outputs

```

boxes: None
keypoints: None
keys: ['probs']
masks: None
names: {0: 'NG_XX ', 1: 'NG_XXX ', 2: 'NG_XXX ', 3: 'NG_XXX ', 4: 'OK'}
orig_img: array([[[ 50, 101,  97],
        [ 50,  96,  97],
        [ 69,  95, 101],
        ...,
        [145, 172, 162],
        [135, 165, 154],
        [134, 163, 154]],

       [[ 47,  95,  89],
        [ 54,  96,  95],
        [ 69,  93,  99],
        ...,
        [142, 171, 162],
        [141, 169, 163],
        [141, 169, 163]],

       [[ 54,  95,  90],
        [ 55,  91,  91],
        [ 69,  89,  94],
        ...,
        [140, 170, 165],
        [140, 170, 165],
        [142, 171, 168]],

       ...,

       [[170, 194, 170],
        [171, 196, 176],
        [165, 191, 175],
        ...,
        [162, 189, 180],
        [159, 186, 176],
        [159, 187, 174]],

       [[162, 192, 163],
        [155, 183, 154],
        [154, 181, 155],
        ...,
        [160, 184, 176],
        [154, 188, 178],
        [151, 189, 177]],

       [[161, 191, 162],
        [158, 186, 157],
        [155, 182, 156],
        ...,
        [156, 181, 171],
        [153, 187, 176],
        [153, 191, 179]]], dtype=uint8)
orig_shape: (65, 67)
path: 'image0.jpg'
probs: array([ 4.1246e-05,     0.68457,  7.5936e-05,  2.7537e-05,  4.2319e-05,  0.00011492,    0.053192,  0.00028872,  0.00020897,  9.0301e-05,  0.00026274,   0.0019836,  8.1003e-05,  6.2883e-05,  0.00011706,   0.0095215,  0.00010473,   0.0013714,     0.24792], dtype=float16)
speed: {'preprocess': None, 'inference': None, 'postprocess': None}
```

0: 224x224 OK 0.61, NG_XX 0.18, NG_XXX 0.09, NG_XXX 0.07, NG_XXX 0.02, 5.2ms
Speed: 0.0ms preprocess, 5.2ms inference, 0.0ms postprocess per image at shape (1, 3, 224, 224)

代码如何获取 TOP1 名称和分数
OK 
0.61

**Comment by monkeycc** [comment]
Thank you for your reply


I need to predict the results
Need to know the predicted results and scores

What should I do

```
0: 224x224    OK 0.61,   NG_XX 0.18, NG_XXX 0.09, NG_XXX 0.07, NG_XXX 0.02, 5.2ms
Speed: 0.0ms preprocess, 5.2ms inference, 0.0ms postprocess per image at shape (1, 3, 224, 224)
```

prediction result 
How to obtain results
`OK   0.61`
@AyushExel

**Comment by AyushExel** [comment]
`result[0].probs` should give you the probability score of at particular Result. See usage here - https://docs.ultralytics.com/modes/predict/#plotting-results

**Comment by monkeycc** [comment]
result[0].probs
```
tensor([3.2755e-04, 1.0638e-01, 1.5710e-03, 6.9446e-04, 5.0282e-04, 3.7475e-04, 1.5485e-02, 1.3358e-03, 7.3719e-02, 7.9577e-02, 1.9133e-03, 2.3774e-03, 3.0329e-04, 1.2074e-03, 7.5453e-04, 3.7560e-03, 2.4922e-03, 1.4962e-02, 6.9227e-01], device='cuda:0')
```


result[0].probs.cpu().numpy()
```
[ 0.00032755     0.10638    0.001571  0.00069446  0.00050282  0.00037475    0.015485   0.0013358    0.073719    0.079577   0.0019133   0.0023774  0.00030329   0.0012074  0.00075453    0.003756   0.0024922    0.014962     0.69227]

```
result

No name
No name ID
Only these

**Comment by glenn-jocher** [comment]
@monkeycc to retrieve the class label and corresponding score from the result of a prediction using YOLOv8, you can access the `result[0].pred` attribute, which contains information for all detected objects in the image including the class label, bounding box location, and confidence score. You can then use this information to extract the label and score for your desired object. To get the label and score in a more convenient format, you can use the `result.print()` method or convert the tensor to a NumPy array with `.cpu().numpy()`.

**Comment by Laughing-q** [comment]
@monkeycc a simple example to get the score, idx and name for classification.
```python
model = YOLO("yolov8n-cls.pt")
im = cv2.imread("ultralytics/assets/bus.jpg")
results = model.predict(source=im)
for result in results:
    probs = result.probs
    score = probs.max()
    idx = probs.argmax()
    name = model.names[int(idx)]
    print(score, idx, name)
```

**Comment by monkeycc** [comment]
Thank you for your reply


I hope to add it to the tutorial documentation
Help more people in need

**Comment by glenn-jocher** [comment]
@monkeycc you're welcome! We're always happy to help. We will consider adding a section to the tutorial documentation covering how to retrieve the predicted class label and score from the `result` object in YOLOv8. Thank you for bringing this to our attention and for your suggestion to improve the documentation. Please feel free to reach out if you have any more questions or if there is anything else we can assist you with.

**Comment by TeenaJO** [comment]
@glenn-jocher 
from ultralytics import YOLO
model = YOLO("/home/ubuntu/Documents/COLOR_MODEL_WEIGHTS/shapecolor/weights/best.pt")
source='/home/ubuntu/Documents/just_one/all1/letters/data/train/test'
results = model.predict(source,save=True)
for r in results:
        for c in range(0,1):
                id=r.probs.top5[c]

                print(id,r.names[id])

this is printing the topmost confidence on my terminal 
Now  want the topmost(highest confidence to be printed on the image) when i give save=True
![augmented_2_vertical_flip_DJI_0026_JPG](https://github.com/ultralytics/ultralytics/assets/129892610/9e6a67a8-58c4-44e4-ac7e-b09370ef7213)
so for the above image,i just want shape_blue to be printed

**Comment by glenn-jocher** [comment]
@TeenaJO thank you for your question. To achieve this in YOLOv8, you could modify the post-processing function that handles the display of the results. After predicting each image, the code identifies the class with the highest confidence from the result object, and then use this name when printing onto the image. By default, `save=True` will save the result image with all class labels for every detected object. To only print the label with the highest confidence, you would need to modify this functionality in the source code. So basically, you'd look into the part of the code where annotations are added to the images and adjust accordingly.

**Comment by TeenaJO** [comment]
@glenn-jocher ,thanks a lot for your reply
i have one more doubt
i am cropping the ROI using the bouonding box coordinates and not with
save_crop=True
        results_model1 = model1(image_path, project=output_folder1,
name='test1', exist_ok=True,  save=False)

        # Extract bounding boxes
        boxes = results_model1[0].boxes.xyxy.tolist()

        # Iterate through the bounding boxes
        for i, box in enumerate(boxes):
            x1, y1, x2, y2 = box

            # Crop the object using the bounding box coordinates
            ultralytics_crop_object = img[int(y1):int(y2), int(x1):int(x2)]
            # Save the cropped object as an image in the output folder
            output_file = os.path.join(output_folder2, 'class_name' +
str(i) + '.jpg')
            cv2.imwrite(output_file, ultralytics_crop_object)
            # Print the folder name, original file name, and cropped image
name
            print(f"Original file: {image_file}")

now i want to get the class_name and the class_id of that particular
object which is being cropped using my above code .Can u pls help me out

**Comment by glenn-jocher** [comment]
@TeenaJO to obtain the class name and id from your result object in the code you are sharing, you can access the 'result[0].classes' attribute which provides information about all the objects found in the current image, including their class identifiers. Specifically, `result[0].classes` will give you a list of class identifiers corresponding to each bounding box in your `boxes` variable. 

Similar to the `.classes` attribute, you can get the class names of your detected objects from the `names` dictionary in your `result`. By finding the maximum index (i.e., the index of the highest scoring class) from `result[0].probs`, you will receive the id of the highest scoring class. Subsequently, you can access the respective class name present in the `names` dictionary of your `result` object using this class id. 

In other words, you can retrieve the class id and corresponding class name for each cropped object within the loop where you are processing each bounding box.

**Comment by TeenaJO** [comment]
i tried using result0].classes but it throwed me a error telling that there
is no attribute like that so i tried result[0].names but its giving me all
the class names .But i want the class name of the object being cropped

        results_model1 = model1(image_path, project=output_folder1,
name='test1', exist_ok=True,  save=False)

        # Load the original image
        img = cv2.imread(image_path)

        # Extract bounding boxes
        boxes = results_model1[0].boxes.xyxy.tolist()

        # Iterate through the bounding boxes
        for i, box in enumerate(boxes):
            x1, y1, x2, y2 = box

            # Crop the object using the bounding box coordinates
            ultralytics_crop_object = img[int(y1):int(y2), int(x1):int(x2)]
            # Save the cropped object as an image in the output folder
            output_file = os.path.join(output_folder2, 'class_name' +
str(i) + '.jpg')
            cv2.imwrite(output_file, ultralytics_crop_object)
            # Print the folder name, original file name, and cropped image
name
            print(f"Original file: {image_file}")
            txt_file.write(f"class_name:{results_model1[0].names}\n")
            # Perform inference on the cropped object with the second model
(shape classification)



Suppose this code is cropping out an object with the boundig box
coordinates ,i want the class name of that object only .

**Comment by glenn-jocher** [comment]
@TeenaJO the `result[0].names` attribute will indeed give you all names of the classes your model can predict. However, in order to get the specific class name for each cropped object within your loop, you will need to access the prediction result for the corresponding bounding box.

This information can be found in `result[0].probs`, which should give a list of probabilities per class for each bounding box, in the same order as your bounding boxes. You can find the index of the highest value in this list, which corresponds to the class id of the dominant class (i.e., the one your model is predicting for this specific bounding box). You can then use this class id to index into `result[0].names`, which will give you the name of that class.

In your loop where you process each bounding box, you can similarly process each corresponding entry in `result[0].probs` to determine the class id and name for the currently considered bounding box. Please ensure that this is done in the same order as your bounding boxes to ensure that the assigned classes indeed correspond to the correct bounding boxes.
